### PR TITLE
expandable_segments <-> other allocator options

### PR DESCRIPTION
**Pull request by zdevito**
Previously setting  garbage_collection_threshold or max_split_size_mb along with expandable_segments:True could cause the allocator to hit assert failures when running nearly out of memory. This PR ensures garbage_collection and max_split freeing do not accidentally try to release expandable segments.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134338

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -2749,7 +2749,8 @@ class DeviceCachingAllocator {
         ? CUDAAllocatorConfig::max_split_size()
         : key.size;
     auto it = pool.blocks.lower_bound(&key);
-    if (it == pool.blocks.end() || (*it)->stream != p.stream()) {
+    if (it == pool.blocks.end() || (*it)->stream != p.stream() ||
+        (*it)->expandable_segment_) {
```
Comment (ezyang) [inline c10/cuda/CUDACachingAllocator.cpp:2753]: Technically, couldn't there be a block that is big enough, but the biggest block is expandable segment and so we picked out the wrong one in the lower_bound test?

Comment (zdevito) [reply c10/cuda/CUDACachingAllocator.cpp:2753]: Yeah, but in that case, it will just go to the else statement and free that single block anyway.

**Review by ezyang — APPROVED** [review]
This looks good.

**Pull request closed without merging**